src/Data/Expr.purs

```diff
@@ -2,89 +2,389 @@ module Lynx.Data.Expr where
 
 import Prelude
 
+import Control.Alt ((<|>))
 import Data.Argonaut (class DecodeJson, class EncodeJson, Json, decodeJson, encodeJson, jsonEmptyObject, stringify, (.:), (:=), (~>))
 import Data.BigInt as Data.BigInt
 import Data.DateTime (DateTime)
 import Data.Either (Either(..))
-import Data.Foldable (fold, intercalate)
+import Data.Eq (class Eq1)
+import Data.Foldable (class Foldable, fold, foldlDefault, foldrDefault)
 import Data.Formatter.DateTime (format, unformat)
 import Data.Formatter.Parser.Interval (extendedDateTimeFormatInUTC)
+import Data.Functor.Coproduct.Inject (inj, prj)
+import Data.Functor.Coproduct.Nested (type (<\/>), (<\/>))
+import Data.Functor.Mu (Mu)
 import Data.Generic.Rep (class Generic)
 import Data.Generic.Rep.Show (genericShow)
-import Data.Maybe (Maybe(..), maybe')
-import Data.Newtype (wrap)
+import Data.Maybe (Maybe(..), maybe, maybe')
+import Data.Newtype (class Newtype, un, wrap)
 import Data.NonEmpty (NonEmpty(..))
 import Data.String.Common (replace)
 import Data.String.Pattern (Pattern(..), Replacement(..))
+import Data.Traversable (class Traversable, traverse)
 import Foreign.Object (Object)
 import Foreign.Object as Foreign.Object
+import Matryoshka (class Corecursive, class Recursive, anaM, cata, embed, project)
 import Ocelot.Data.Currency (Cents, formatCentsToStrDollars)
 import Test.QuickCheck (class Arbitrary, arbitrary)
 import Test.QuickCheck.Gen (Gen, Size, arrayOf, oneOf, sized)
 
 type Key = String
 
-data ExprType
-  = Array (Array ExprType)
-  | Boolean Boolean
-  | Cents Cents
-  | DateTime DateTime
-  | Int Int
-  | Pair { name :: ExprType, value :: ExprType }
-  | String String
+newtype ArrayF a
+  = Array (Array a)
+
+derive instance eqArrayF :: (Eq a) => Eq (ArrayF a)
+
+derive instance eq1ArrayF :: Eq1 ArrayF
+
+derive instance functorArrayF :: Functor ArrayF
+
+derive newtype instance showArrayF :: (Show a) => Show (ArrayF a)
+
+derive newtype instance foldableArrayF :: Foldable ArrayF
+
+derive newtype instance traversableArrayF :: Traversable ArrayF
+
+decodeArrayF :: ExprJSON -> Either String (ArrayF ExprJSON)
+decodeArrayF { out, param } = case out of
+  "Array" -> map Array (decodeJson param)
+  _ -> Left "Expected Array"
+
+encodeArrayF :: ArrayF ExprJSON -> ExprJSON
+encodeArrayF = case _ of
+  Array x -> base "Array" x
```

Comment:
I already commented on `foldMap` below, but noticing this here too, I'm wondering if there's some motivation for using `case` when there's just one constructor? I feel like these seem more concise

```purescript
encodeArrayF (Array x) = base "Array" x
```
```purescript
encodeArrayF = base "Array" <<< unwrap
```

Reply:
Is it just consistency?

Reply:
It's more about dealing with the codebase as it changes over time.

With these Haskell-based languages, top-level pattern matches get pretty noisy. It's harder to tell at a glance what is important and what isn't because the whole value changes. For example, if the `Array` constructor changes, that change is separate from the name of the function or whatever other args we pass to the function.

I don't know that `ArrayF _` is going to change a whole bunch as the code evolves, so it might be fine to do a top-level pattern match. I also don't know that `ArrayF _` is going to stay the same as the code evolves, so it seems safer to have it as a regular case statement.

We can definitely write it as a top-level pattern match if that's what you'd prefer.

Reply:
> With these Haskell-based languages, top-level pattern matches get pretty noisy. It's harder to tell at a glance what is important and what isn't because the whole value changes. For example, if the Array constructor changes, that change is separate from the name of the function or whatever other args we pass to the function.

That's a solid argument for separating the function name from the pattern matching.

> We can definitely write it as a top-level pattern match if that's what you'd prefer.

I don't have a preference. I value conciseness, I also value consistency, and separation of concerns.

Consistency could mean always going for `case`, or it could be a little more nuanced, i.e. pattern matching at the top level when it's only one constructor, casing otherwise. As long as we're consistent. 

I do think when dealing with newtypes it's valid just to use `unwrap`. For example, what about `let` bindings or inline functions?

```purescript
_.teamId <<< unwrap <$> maybeTeam
```
I think is a lot nicer than getting at that value via casing. Of course, it's a different scenario than here, but it's still a question of where to draw the scope of this rule. I might be overthinking.

Anyway, open to having `case` be the standard, was just looking for clarification on why it's your default. 🙂

Reply:
Cool. What should we do here, top-level or case? Either one seems fine, and I'm hapy to change this code before merging.

Reply:
I'm down to keep this and nail down our rules around pattern matching later